File: plugins/rider/src/main/kotlin/com/shortsleevestudio/gamescript/database/SqliteConnection.kt
package com.shortsleevestudio.gamescript.database

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.sql.Connection
import java.sql.DriverManager
import java.sql.PreparedStatement
import java.sql.ResultSet
import java.util.concurrent.atomic.AtomicInteger
import java.util.concurrent.atomic.AtomicReference

/**
 * SQLite database connection using JDBC.
 * Provides query, run, and transaction support.
 */
class SqliteConnection(private val path: String) {

    private var connection: Connection? = null
    private val nextTransactionId = AtomicInteger(1)
    private val activeTransaction = AtomicReference<TransactionInfo?>(null)

    private data class TransactionInfo(
        val id: String,
        val startTime: Long
    )

    /**
     * Connect to the SQLite database.
     */
    suspend fun connect() = withContext(Dispatchers.IO) {
        // Load SQLite JDBC driver
        Class.forName("org.sqlite.JDBC")

        connection = DriverManager.getConnection("jdbc:sqlite:$path")
        connection?.autoCommit = true

        // Enable WAL mode for better concurrency
        exec("PRAGMA journal_mode=WAL")
        // Set busy timeout to 30 seconds
        exec("PRAGMA busy_timeout=30000")
    }

    /**
     * Close the connection.
     */
    fun close() {
        try {
            connection?.close()
            connection = null
        } catch (_: Exception) {
            // Ignore close errors
        }
    }

    /**
     * Execute a SELECT query.
     */
    suspend fun query(sql: String, params: List<Any?>? = null): List<Map<String, Any?>> =
        withContext(Dispatchers.IO) {
            val conn = connection ?: throw IllegalStateException("Not connected")

            conn.prepareStatement(sql).use { stmt ->
                bindParameters(stmt, params)
                stmt.executeQuery().use { rs ->
                    resultSetToList(rs)
                }
            }
        }

    /**
     * Execute an INSERT/UPDATE/DELETE statement.
     */
    suspend fun run(sql: String, params: List<Any?>? = null, returning: Boolean = false): RunResult =
        withContext(Dispatchers.IO) {
            val conn = connection ?: throw IllegalStateException("Not connected")

            if (returning) {
                // For RETURNING clause, use executeQuery
                conn.prepareStatement(sql).use { stmt ->
                    bindParameters(stmt, params)
                    stmt.executeQuery().use { rs ->
                        val rows = resultSetToList(rs)
                        val lastId = rows.firstOrNull()?.get("id") as? Long
                        RunResult(lastId, rows.size.toLong(), rows)
                    }
                }
            } else {
                conn.prepareStatement(sql, PreparedStatement.RETURN_GENERATED_KEYS).use { stmt ->
                    bindParameters(stmt, params)
                    val changes = stmt.executeUpdate()

                    val lastId = stmt.generatedKeys.use { keys ->
                        if (keys.next()) keys.getLong(1) else null
                    }

                    RunResult(lastId, changes.toLong(), emptyList())
                }
            }
        }

    /**
     * Execute a DDL statement.
     */
    suspend fun exec(sql: String) = withContext(Dispatchers.IO) {
        val conn = connection ?: throw IllegalStateException("Not connected")
        conn.createStatement().use { stmt ->
            stmt.execute(sql)
        }
    }

    /**
     * Begin a new transaction.
     */
    suspend fun beginTransaction(): String = withContext(Dispatchers.IO) {
        check(activeTransaction.get() == null) {
            "SQLite does not support nested transactions"
        }

        val id = "tx_${nextTransactionId.getAndIncrement()}"
        exec("BEGIN IMMEDIATE")
        activeTransaction.set(TransactionInfo(id, System.currentTimeMillis()))
        id
    }

    /**
     * Commit a transaction.
     */
    suspend fun commitTransaction(connectionId: String) = withContext(Dispatchers.IO) {
        val tx = activeTransaction.get()
        check(tx?.id == connectionId) { "Transaction $connectionId not found" }

        exec("COMMIT")
        activeTransaction.set(null)
    }

    /**
     * Rollback a transaction.
     */
    suspend fun rollbackTransaction(connectionId: String) = withContext(Dispatchers.IO) {
        val tx = activeTransaction.get()
        check(tx?.id == connectionId) { "Transaction $connectionId not found" }

        exec("ROLLBACK")
        activeTransaction.set(null)
    }

    private fun bindParameters(stmt: PreparedStatement, params: List<Any?>?) {
        params?.forEachIndexed { index, value ->
            when (value) {
                null -> stmt.setNull(index + 1, java.sql.Types.NULL)
                is String -> stmt.setString(index + 1, value)
                is Int -> stmt.setInt(index + 1, value)
                is Long -> stmt.setLong(index + 1, value)
                is Double -> stmt.setDouble(index + 1, value)
                is Float -> stmt.setFloat(index + 1, value)
                is Boolean -> stmt.setBoolean(index + 1, value)
                is ByteArray -> stmt.setBytes(index + 1, value)
                else -> stmt.setObject(index + 1, value)
            }
        }
    }

    private fun resultSetToList(rs: ResultSet): List<Map<String, Any?>> {
        val metaData = rs.metaData
        val columnCount = metaData.columnCount
        val columns = (1..columnCount).map { metaData.getColumnLabel(it) }

        val results = mutableListOf<Map<String, Any?>>()
        while (rs.next()) {
            val row = mutableMapOf<String, Any?>()
            columns.forEachIndexed { index, column ->
                row[column] = rs.getObject(index + 1)
            }
            results.add(row)
        }
        return results
    }
}

File: plugins/dotnet/visual-studio/GameScript.VisualStudio/Database/SqliteConnection.cs
using Microsoft.Data.Sqlite;

namespace GameScript.VisualStudio;

/// <summary>
/// SQLite database connection using Microsoft.Data.Sqlite.
/// Provides query, run, and transaction support with WAL mode.
/// </summary>
public sealed class SqliteConnection : IDisposable
{
    private readonly string _path;
    private Microsoft.Data.Sqlite.SqliteConnection? _connection;
    private int _nextTransactionId = 1;
    private TransactionInfo? _activeTransaction;
    private readonly object _transactionLock = new();

    private sealed class TransactionInfo
    {
        public required string Id { get; init; }
        public required SqliteTransaction Transaction { get; init; }
    }

    public SqliteConnection(string path)
    {
        _path = path;
    }

    /// <summary>
    /// Connect to the SQLite database.
    /// </summary>
    public async Task ConnectAsync(CancellationToken cancellationToken = default)
    {
        string connectionString = new SqliteConnectionStringBuilder
        {
            DataSource = _path,
            Mode = SqliteOpenMode.ReadWriteCreate,
            Cache = SqliteCacheMode.Shared
        }.ToString();

        _connection = new Microsoft.Data.Sqlite.SqliteConnection(connectionString);
        await _connection.OpenAsync(cancellationToken);

        // Enable WAL mode for better concurrency
        await ExecAsync("PRAGMA journal_mode=WAL", cancellationToken);
        // Set busy timeout to 30 seconds
        await ExecAsync("PRAGMA busy_timeout=30000", cancellationToken);
    }

    /// <summary>
    /// Close the connection.
    /// </summary>
    public async Task CloseAsync()
    {
        if (_connection != null)
        {
            await _connection.CloseAsync();
            await _connection.DisposeAsync();
            _connection = null;
        }
    }

    /// <summary>
    /// Execute a SELECT query.
    /// </summary>
    public async Task<List<Dictionary<string, object?>>> QueryAsync(
        string sql,
        List<object?>? parameters = null,
        CancellationToken cancellationToken = default)
    {
        EnsureConnected();

        await using SqliteCommand cmd = _connection!.CreateCommand();
        cmd.CommandText = sql;
        BindParameters(cmd, parameters);

        await using SqliteDataReader reader = await cmd.ExecuteReaderAsync(cancellationToken);
        return await ReaderToListAsync(reader, cancellationToken);
    }

    /// <summary>
    /// Execute an INSERT/UPDATE/DELETE statement.
    /// </summary>
    public async Task<RunResult> RunAsync(
        string sql,
        List<object?>? parameters = null,
        bool returning = false,
        CancellationToken cancellationToken = default)
    {
        EnsureConnected();

        await using SqliteCommand cmd = _connection!.CreateCommand();
        cmd.CommandText = sql;
        BindParameters(cmd, parameters);

        if (returning)
        {
            // For RETURNING clause, use ExecuteReader
            await using SqliteDataReader reader = await cmd.ExecuteReaderAsync(cancellationToken);
            List<Dictionary<string, object?>> rows = await ReaderToListAsync(reader, cancellationToken);
            Dictionary<string, object?>? firstRow = rows.FirstOrDefault();
            long? lastId = firstRow?.TryGetValue("id", out object? idValue) == true ? Convert.ToInt64(idValue) : null;
            return new RunResult
            {
                LastId = lastId,
                Changes = rows.Count,
                Rows = rows
            };
        }
        else
        {
            int changes = await cmd.ExecuteNonQueryAsync(cancellationToken);

            // Get last insert rowid
            await using SqliteCommand lastIdCmd = _connection.CreateCommand();
            lastIdCmd.CommandText = "SELECT last_insert_rowid()";
            object? lastIdResult = await lastIdCmd.ExecuteScalarAsync(cancellationToken);

            return new RunResult
            {
                LastId = lastIdResult != null && lastIdResult != DBNull.Value ? Convert.ToInt64(lastIdResult) : null,
                Changes = changes,
                Rows = new List<Dictionary<string, object?>>()
            };
        }
    }

    /// <summary>
    /// Execute a DDL statement.
    /// </summary>
    public async Task ExecAsync(string sql, CancellationToken cancellationToken = default)
    {
        EnsureConnected();

        await using SqliteCommand cmd = _connection!.CreateCommand();
        cmd.CommandText = sql;
        await cmd.ExecuteNonQueryAsync(cancellationToken);
    }

    /// <summary>
    /// Begin a new transaction.
    /// </summary>
    public Task<string> BeginTransactionAsync(CancellationToken cancellationToken = default)
    {
        EnsureConnected();

        lock (_transactionLock)
        {
            if (_activeTransaction != null)
            {
                throw new InvalidOperationException("SQLite does not support nested transactions");
            }

            string id = $"tx_{Interlocked.Increment(ref _nextTransactionId)}";

            // BeginTransaction with deferred=false acquires an immediate write lock
            SqliteTransaction transaction = _connection!.BeginTransaction(deferred: false);

            _activeTransaction = new TransactionInfo { Id = id, Transaction = transaction };

            return Task.FromResult(id);
        }
    }

    /// <summary>
    /// Commit a transaction.
    /// </summary>
    public async Task CommitTransactionAsync(string connectionId, CancellationToken cancellationToken = default)
    {
        TransactionInfo? tx;
        lock (_transactionLock)
        {
            tx = _activeTransaction;
            if (tx?.Id != connectionId)
            {
                throw new InvalidOperationException($"Transaction {connectionId} not found");
            }
        }

        await tx.Transaction.CommitAsync(cancellationToken);

        lock (_transactionLock)
        {
            _activeTransaction = null;
        }
    }

    /// <summary>
    /// Rollback a transaction.
    /// </summary>
    public async Task RollbackTransactionAsync(string connectionId, CancellationToken cancellationToken = default)
    {
        TransactionInfo? tx;
        lock (_transactionLock)
        {
            tx = _activeTransaction;
            if (tx?.Id != connectionId)
            {
                throw new InvalidOperationException($"Transaction {connectionId} not found");
            }
        }

        await tx.Transaction.RollbackAsync(cancellationToken);

        lock (_transactionLock)
        {
            _activeTransaction = null;
        }
    }

    private void BindParameters(SqliteCommand cmd, List<object?>? parameters)
    {
        if (parameters == null) return;

        for (int i = 0; i < parameters.Count; i++)
        {
            object? value = parameters[i];
            SqliteParameter param = cmd.CreateParameter();
            param.ParameterName = $"${i + 1}";
            param.Value = value ?? DBNull.Value;
            cmd.Parameters.Add(param);
        }
    }

    private static async Task<List<Dictionary<string, object?>>> ReaderToListAsync(
        SqliteDataReader reader,
        CancellationToken cancellationToken)
    {
        List<Dictionary<string, object?>> results = new();

        while (await reader.ReadAsync(cancellationToken))
        {
            Dictionary<string, object?> row = new();
            for (int i = 0; i < reader.FieldCount; i++)
            {
                string columnName = reader.GetName(i);
                object? value = reader.IsDBNull(i) ? null : reader.GetValue(i);
                row[columnName] = value;
            }
            results.Add(row);
        }

        return results;
    }

    private void EnsureConnected()
    {
        if (_connection == null || _connection.State != System.Data.ConnectionState.Open)
        {
            throw new InvalidOperationException("Not connected");
        }
    }

    public void Dispose()
    {
        _activeTransaction?.Transaction.Dispose();
        _connection?.Dispose();
    }
}
